README.md
# Diagram As Text

The tool to generate diagram based on textual description. Although it takes quite some effort to make a diagram, LLM is
here for the rescue! 🤖🦾

<video src="https://user-images.githubusercontent.com/13434797/218483498-57d671ad-6ad0-40af-b107-7d4cb87cc673.mp4" controls="controls" style="max-width: 720px;"></video>

🚀🚀🚀 _Mission_: to enable anyone to explain complex system, or process in a simple way.

🚀🚀 _Objective_: to streamline knowledge sharing through diagrams.

🚀 _MVP_: plain english to [C4 container](http://c4model.com/) diagrams.

🔔 **Wanted**: founding contributors 🔔

The project is purely community driven - we need your support:

- Please give the project a _star_.
- Join us as _contributor_: we need software engineers, data scientists, analysts, designers.

If you are excited about the project, feel comfortable with our [ways of work](#ways-of-work), and want to contribute,
please [get in touch](#contacts) for further details.

Thank you! 🙏

## Contacts

- Submit
  your [request](https://github.com/kislerdm/diagramastext/issues/new?assignee=kislerdm&labels=feedback&title=PLEASE%20DEFINE%20YOUR%20REQUEST&body=%23%23%20What%0APlease%20describe%20your%20proposal.%0A%0A%23%23%20Why%0APlease%20clarify%20the%20context.%0A%0A%23%23%20How%0A%0A(optional)%20Please%20sketch%20execution%20paths.)
- Join us on [Slack](https://join.slack.com/t/diagramastextdev/shared_invite/zt-1onedpbsz-ECNIfwjIj02xzBjWNGOllg)
- Write us: <a href="mailto:hi@diagramastext.com">hi@diagramastext.com</a>
- Get in touch on [LinkedIn](https://www.linkedin.com/in/dkisler/)

<details>
<summary><strong>Outlook</strong></summary>

* [Ways of work](#ways-of-work)
    + [Manifesto](#manifesto)
    + [Process](#process)
    + [Bets/Panning/Commitment](#betspanningcommitment)
* [Tech details](#tech-details)
    + [Architecture](#architecture)
    + [Local setup](#local-setup)
* [License](#license)
    + [Codebase](#codebase)
    + [Images and diagrams](#images-and-diagrams)

</details>

## Ways of work

### Manifesto

- We are driven by the mission
- We respect one another and the community
- We deliver in lean iterations
- We work async with pairing programming sessions
- We share the work openly, see the [license details](#license)

### Process

- We follow [TDD](https://www.guru99.com/test-driven-development.html)
- We follow [RDD](https://tom.preston-werner.com/2010/08/23/readme-driven-development.html)
- We maintain flat modules structure whether possible
- We embrace clean architecture inspired by Alistair
  Cockburn's [port-adapter approach](https://web.archive.org/web/20180822100852/http://alistair.cockburn.us/Hexagonal+architecture)
  and Martin Fowler's [inversion of control](https://martinfowler.com/bliki/InversionOfControl.html), see a great
  write-up [here](https://jmgarridopaz.github.io/content/hexagonalarchitecture.html#tc2-3) on the subjects matter
- We aim for simplicity with the least external dependencies
- We follow conventional _comments_ [guideline](https://conventionalcomments.org/) for code reviews
- We follow conventional _commits_ [guideline](https://www.conventionalcommits.org/en/v1.0.0/):
    - `feat`: for features
    - `fix`: for defect fix
    - `chore`: for infra, ci, or docs adjustments; or refactoring
- We follow the [monorepo](https://monorepo.tools/) approach
- We follow [trunk-based development](https://trunkbaseddevelopment.com/) model
- We follow the release [guideline](https://keepachangelog.com/en/1.0.0/) and [semantic versioning](https://semver.org/)

### Bets/Panning/Commitment

- [Issues Board](https://github.com/users/kislerdm/projects/5/views/)

## Tech details

<details>
<summary><strong>Tech stack</strong></summary>

- Languages:
    - Go 1.19
    - TypeScript, ES2020:
      - [vite](https://vitejs.dev)
      - [vitest](https://vitest.dev/)
    - Python 3.9
- Markup:
    - Markdown
    - HTML5
    - CSS3
- CI:
    - GitHub Actions
- Infra:
    - GCP
        - IAM
        - SecretsManager
        - CloudRun
        - IdentityPlatform
    - AWS (infra deprecated as of 2023-03)
    - GitHub Pages
    - [Neon](https://neon.tech/)
    - Cloudflare
    - namecheap
    - godaddy
- Tools:
    - gnuMake
    - Docker
    - terraform
- Logic:
    - PlantUML
    - OpenAI

</details>

### Architecture

#### C4 Context Level

![architecture-context](architecture_c4_l1.svg)

### Local setup

#### Requirements

- git
- gnuMake
- docker ~> 20.10

#### Steps

Follow the steps to provision local development environment.

1. Clone the repo

```commandline
git clone git@github.com:kislerdm/diagramastext.git
```

2. Export OpenAI access [token](https://platform.openai.com/docs/api-reference/authentication) as the environment
   variable `OPENAI_API_KEY`:

```commandline
export OPENAI_API_KEY=##TOKEN##
```

3. Run to provision the local environment:

```commandline
make localenv
```

**Note**:
- The webclient will be available on http://localhost:9001
- The database webclient will be available on http://localhost:9081

4. Run to clean the local environment:

```commandline
make localenv-teardown
```

## License

### Codebase

The codebase is distributed under the [Apache 2.0 licence](LICENSE).

### Images and diagrams

<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/80x15.png" /></a><br />
This work
by <a xmlns:cc="http://creativecommons.org/ns#" href="diagramastext.dev" property="cc:attributionName" rel="cc:attributionURL">
diagramastext.dev</a> is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">
Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.
Based on a work at [https://github.com/kislerdm/diagramastext](https://github.com/kislerdm/diagramastext).
Permissions beyond the scope of this license may be available
at <a xmlns:cc="http://creativecommons.org/ns#" href="diagramastext.dev" rel="cc:morePermissions">diagramastext.dev</a>. 


CHANGELOG.md
# Changelog

## [0.0.7] - 2023-07-18

### Fixed

- Deserialization of the OpenAI response

## [0.0.6] - 2023-04-13

### Added

- Authentication and authorization layer to use the Rest API.
- `/quotas` endpoint to fetch the service usage

### Fixed

- CORS: validation of preflight calls.
- Response's content-type header value.

### Changed

- OpenAI's tokens consumption tracking.

## [0.0.6beta] - 2023-04-12

### Added

- Rest API interface to the core logic. Find the specification [here](https://diagramastext.dev/api-reference/).

### Fixed

- OpenAI's prediction parsing and deserialization when the graph's JSON is surrounded by text.

**_Example_**

_GIVEN_: The OpenAI chat response's content below  

> Here's the C4 diagram for a Python web server reading from an external Postgres database:
> 
> ```
> {
>   "title": "Python Web Server Reading from External Postgres Database",
>   "nodes": [
>     {"id": "0", "label": "Web Server", "technology": "Python"},
>     {"id": "1", "label": "Postgres", "technology": "Postgres", "external": true, "is_database": true}
>   ],
>   "links": [
>     {"from": "0", "to": "1", "label": "reads from Postgres", "technology": "TCP"}
>   ],
>   "footer": "C4 Model"
> }
> ```
> 
> The diagram shows two nodes: a Python web server and an external Postgres database. The web server reads data from the Postgres database over TCP.

_WHEN_: apply new deserialization and parsing logic

_THEN_: get the following graph definition:

```
{"title":"Python Web Server Reading from External Postgres Database","nodes":[{"id":"0","label":"Web Server","technology":"Python"},{"id":"1","label":"Postgres","technology":"Postgres","external":true,"is_database":true}],"links":[{"from":"0","to":"1","label":"reads from Postgres","technology":"TCP"}],"footer":"C4 Model"}
```

### Changed

- Migrated to [pgx](https://github.com/jackc/pgx) from [lib/pq](https://github.com/lib/pq). 

## [0.0.5] - 2023-04-05

### Added

- User input validation according to the user status, i.e. registered/non-registered split
- Improved user feedback in case the model fails to predict a graph
- Trigger diagram generation process on the ctrl+enter keys press
- Cancellation of requests over 10 sec. on the escape key press
- Hide popup on the escape key press

### Changed

- Refactored the webclient codebase:
  - typescript
  - vite
  - tests coverage improvement

## [0.0.4] - 2023-03-23

### Added

- Support for the [`User`](https://github.com/plantuml-stdlib/C4-PlantUML/#supported-diagram-types) C4 macro.
- Diagram legend by default. It can be removed upon an explicit prompt's request only. 
**Examples**: 

The prompt yields the diagram with the legend: 
```
diagram of three boxes 
```

The prompts yield the diagram without the legend:

```
diagram of three boxes without legend
```

```
diagram of three boxes, no legend
```

```
diagram of three boxes with no legend
```

```
diagram of three boxes. remove legend
```

### Fixed

- ([#56](https://github.com/kislerdm/diagramastext/issues/56)) Moved from the model
  from "[code-davinci-002](https://platform.openai.com/docs/models/codex)"
  to "[gpt-3.5-turbo](https://platform.openai.com/docs/models/gpt-3-5)" following on the OpenAI announcement to
  discontinue support for the Codex API.

## [0.0.3] - 2023-03-21

### Fixed

- Migrated the core logic to GCP to bypass the OpenAI latency exceeding the AWS GW request limit of 29 sec. _Note_ that
  the tf codebase for AWS infra is preserved for the release.

### Changed

- Refactored the [`core`](./server/core) logic:
    - [Port-adaptor](https://web.archive.org/web/20180822100852/http://alistair.cockburn.us/Hexagonal+architecture)
      approach
    - Split sub-packages to dedicated modules in the `pkg` directory. Note that the packages below released as `v0.0.1`:
        - HTTP client with the backoff-retry mechanism
        - GCP Secretsmanager
        - Postgres client
        - OpenAI client
- [tf](infrastructure/neon) state with the Neon db provisioning migrated to GCP.

## [0.0.2] - 2023-02-21

### Changed

- Model configurations were changed:

```json
{
  "model": "code-davinci-002",
  "max_tokens": 768,
  "temperature": 0.2,
  "best_of": 3,
  "top_p": 1,
  "frequency_penalty": 0,
  "presence_penalty": 0,
  "stop": [
    "\n"
  ]
}
```

- (minor) Added the logo

### Security

- Added the vault to manage access credentials to improve security

## [0.0.1] - 2023-02-13

Beta release.

### Added

- Core server-side logic to deliver [C4 model](https://c4model.com/) diagrams
    - OpenAI agent
    - PlantUML agent
    - AWS Lambda interface
    - HTTP interface
- Webclient:
    - Freetext intput with the input validation
    - Output
    - Go and download buttons

#### Open IA Model

Configuration:

```json
{
  "model": "code-cushman-001",
  "max_tokens": 768,
  "temperature": 0.2,
  "best_of": 3,
  "top_p": 1,
  "frequency_penalty": 0,
  "presence_penalty": 0,
  "stop": [
    "\n"
  ]
}
```


core/openai.go
package core

import (
	"bytes"
	"context"
	_ "embed"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"
)

/*
	Defines client to communicate to OpenAI over http
*/

// ConfigOpenAI configuration of the OpenAI client.
// see:
//   - https://platform.openai.com/docs/api-reference/authentication
//   - https://platform.openai.com/docs/api-reference/completions
type ConfigOpenAI struct {
	// https://platform.openai.com/docs/api-reference/authentication
	Token string

	// https://platform.openai.com/docs/api-reference/requesting-organization
	Organization string

	// https://platform.openai.com/docs/api-reference/completions/create#completions/create-model
	Model string

	// https://platform.openai.com/docs/api-reference/completions/create#completions/create-max_tokens
	MaxTokens int

	// https://platform.openai.com/docs/api-reference/completions/create#completions/create-temperature
	Temperature float32
}

type openAIRequest struct {
	Model            string   `json:"model"`
	Prompt           string   `json:"prompt"`
	Stop             []string `json:"stop,omitempty"`
	MaxTokens        int      `json:"max_tokens,omitempty"`
	Temperature      float32  `json:"temperature,omitempty"`
	TopP             float32  `json:"top_p"`
	FrequencyPenalty float32  `json:"frequency_penalty"`
	PresencePenalty  float32  `json:"presence_penalty"`
	BestOf           int8     `json:"best_of"`
}

type clientOpenAI struct {
	httpClient   HttpClient
	payload      openAIRequest
	sink         func(s string)
	token        string
	organization string
	baseURL      string
}

// NewOpenAIClient initiates the client to communicate with the plantuml server.
func NewOpenAIClient(cfg ConfigOpenAI, optFns ...func(client *clientOpenAI)) (ClientInputToGraph, error) {
	c := clientOpenAI{
		httpClient:   nil,
		token:        cfg.Token,
		organization: cfg.Organization,
		payload: openAIRequest{
			Model:            cfg.Model,
			Prompt:           "",
			Stop:             []string{"\n"},
			MaxTokens:        cfg.MaxTokens,
			TopP:             defaultTopP,
			Temperature:      cfg.Temperature,
			FrequencyPenalty: 0,
			PresencePenalty:  0,
			BestOf:           defaultBestOf,
		},
		baseURL: baseURLOpenAI,
		sink:    nil,
	}

	for _, fn := range optFns {
		fn(&c)
	}

	if err := resolveConfigurations(&c); err != nil {
		return nil, Error{
			Service:                   ServiceOpenAI,
			Stage:                     StageInit,
			ServiceResponseStatusCode: 0,
			Message:                   err.Error(),
		}
	}

	resolveHTTPClientOpenAI(&c)

	return &c, nil
}

// WithHTTPClientOpenAI overwrite the OpenAI HTTP client.
func WithHTTPClientOpenAI(c HttpClient) func(o *clientOpenAI) {
	return func(o *clientOpenAI) {
		o.httpClient = c
	}
}

// WithSinkFn sets sink function to record model's input and output.
func WithSinkFn(sinkFn func(string)) func(o *clientOpenAI) {
	return func(o *clientOpenAI) {
		o.sink = sinkFn
	}
}

func resolveConfigurations(c *clientOpenAI) error {
	if c.token == "" {
		return errors.New(
			"'Token' must be specified, see: https://platform.openai.com/docs/api-reference/authentication",
		)
	}

	if c.payload.Model == "" {
		c.payload.Model = defaultModelOpenAI
	}

	if c.payload.MaxTokens <= 0 || c.payload.MaxTokens > 2048 {
		c.payload.MaxTokens = defaultMaxTokens
	}

	if c.payload.Temperature <= 0 || c.payload.Temperature > 1 {
		c.payload.Temperature = defaultTemperature
	}

	return nil
}

func resolveHTTPClientOpenAI(c *clientOpenAI) {
	if c.httpClient == nil {
		c.httpClient = &http.Client{Timeout: defaultTimeoutOpenAI}
	}
}

const (
	baseURLOpenAI        = "https://api.openai.com/v1/"
	defaultTimeoutOpenAI = 3 * time.Minute
	defaultModelOpenAI   = "code-cushman-001"
	defaultMaxTokens     = 768
	defaultTemperature   = 0.2
	defaultTopP          = 1
	defaultBestOf        = 3
)

type openAIResponse struct {
	ID      string `json:"id"`
	Object  string `json:"object"`
	Created int    `json:"created"`
	Model   string `json:"model"`
	Choices []struct {
		Text         string `json:"text"`
		Index        int    `json:"index"`
		Logprobs     int    `json:"logprobs"`
		FinishReason string `json:"finish_reason"`
	} `json:"choices"`
	Usage struct {
		PromptTokens     int `json:"prompt_tokens"`
		CompletionTokens int `json:"completion_tokens"`
		TotalTokens      int `json:"total_tokens"`
	} `json:"usage"`
}

func (c *clientOpenAI) Do(ctx context.Context, prompt string) (DiagramGraph, error) {
	// FIXME: revert back for test
	// FIXME: replace with embed after understanding encoding diff:
	// FIXME: we tried to replace with embedded files already, but openAI was returning 400
	// FIXME: it's likely to do with data encoding -> TBD
	const promptInputComment = `Given prompts and corresponding graphs as json define new graph based on new prompt. ` +
		`Every node has id,label,group,technology as strings,external,is_queue and is_database as bool. ` +
		`Every link connects nodes using their id:from,to. It also has label,technology and direction as strings. ` +
		`Every json has title and footer as string.` + "\n" +
		`Draw c4 container diagram with four containers,thee of which are external and belong to the system X.
	{"nodes":[{"id":"0"},{"id":"1","group":"X","external":true},{"id":"2","group":"X","external":true},` +
		`{"id":"3","group":"X","external":true}]}` + "\n" +
		`three connected boxes
	{"nodes":[{"id":"0"},{"id":"1"},{"id":"2"}],` +
		`"links":[{"from":"0","to":"1"},{"from":"1","to":"2"},{"from":"2","to":"0"}]}` + "\n" +
		`c4 containers:golang web server authenticating users read from external mysql database
	{"nodes":[{"id":"0","label":"Web Server","technology":"Go","description":"Authenticates users"},` + "\n" +
		`{"id":"1","label":"Database","technology":"MySQL","external":true,"is_database":true}]` + "\n" +
		`"links":[{"from":"0","to":"1","direction":"LR"}]}` + "\n" +
		`Five containers in three groups. First container is a Flink Application which performs feature engineering ` +
		`using JSON encoded user behavioural clickstream consumed from AWS Kinesis Stream over HTTP. ` +
		`It publishes AVRO encoded results to the kafka topic over TCP and infers the machine learning model by ` +
		`sending JSON data over rest API. The Flink application is deployed to AWS KDA of the Business Domain account. ` +
		`Kafka topic is part of the Streaming Platform,which sinks the data to the Datalake,AWS S3 bucket. ` +
		`The model is deployed to the MLPlatform. MLPlatform,clickstream and datalake belong to the Data Platform. ` +
		`All but Flink application are external.` + "\n" +
		`{"nodes":[{"id":"0","label":"Go producer","technology":"Go","description":"Publishes to Kafka"},` +
		`{"id":"1","label":"Kafka","technology":"Kafka","is_queue":true},` +
		`{"id":"2","label":"Java consumer","technology":"Java","description":"Consumes from Kafka"}],` +
		`"links":[{"from":"0","to":"1","label":"publishes to Kafka","technology":"TCP/Protobuf"},` +
		`{"from":"2","to":"1","label":"consumes from Kafka","technology":"TCP/Protobuf"}]}` + "\n" +
		`draw c4 diagram with python backend reading from postgres over tcp` + "\n" +
		`{"nodes":[{"id":"0","label":"Postgres","technology":"Postgres","is_database":true},` +
		`{"id":"1","label":"Backend","technology":"Python"}],` +
		`"links":[{"from":"1","to":"0","label":"reads from postgres","technology":"TCP"}]}` + "\n" +
		`draw c4 diagram with java backend reading from dynamoDB over tcp` + "\n" +
		`{"nodes":[{"id":"0","label":"DynamoDB","technology":"DynamoDB","is_database":true},` +
		`{"id":"1","label":"Backend","technology":"Java"}],` +
		`"links":[{"from":"1","to":"0","label":"reads from dynamoDB","technology":"TCP"}]}` + "\n" +
		`c4 diagram with kotlin backend reading from mysql and publishing to kafka avro encoded events` + "\n" +
		`{"nodes":[{"id":"0","label":"Backend","technology":"Kotlin"},` +
		`{"id":"1","label":"Kafka","technology":"Kafka","is_queue":true},` +
		`{"id":"2","label":"Database","technology":"MySQL","is_database":true}],` +
		`"links":[{"from":"0","to":"2","label":"reads from database","technology":"TCP"},` +
		`{"from":"0","to":"2","label":"publishes to kafka","technology":"TCP/AVRO"}]`
	payload := c.payload
	payload.Prompt = promptInputComment + "\n" + strings.ReplaceAll(prompt, "\n", "") + "\n"

	respBytes, err := c.requestHandler(ctx, payload)
	if err != nil {
		return DiagramGraph{}, Error{
			Service:                   ServiceOpenAI,
			Stage:                     StageRequest,
			Message:                   err.Error(),
			ServiceResponseStatusCode: 0,
		}
	}

	return c.decodeResponse(ctx, respBytes)
}

func (c *clientOpenAI) setHeader(req *http.Request) {
	req.Header.Add("Authorization", "Bearer "+c.token)
	req.Header.Add("Content-Type", "application/json")
	if c.organization != "" {
		req.Header.Add("Organization", c.organization)
	}
}

func (c *clientOpenAI) decodeResponse(ctx context.Context, respBytes []byte) (DiagramGraph, error) {
	var resp openAIResponse
	if err := json.Unmarshal(respBytes, &resp); err != nil {
		return DiagramGraph{}, Error{
			Service:                   ServiceOpenAI,
			Stage:                     StageDeserialization,
			Message:                   err.Error(),
			ServiceResponseStatusCode: 0,
		}
	}

	if len(resp.Choices) == 0 {
		return DiagramGraph{}, Error{
			Service:                   ServiceOpenAI,
			Stage:                     StageDeserialization,
			Message:                   "openAI could not convert the input prompt",
			ServiceResponseStatusCode: 0,
		}
	}

	s := cleanRawResponse(resp.Choices[0].Text)

	var o DiagramGraph
	if err := json.Unmarshal([]byte(s), &o); err != nil {
		return DiagramGraph{}, Error{
			Service:                   ServiceOpenAI,
			Stage:                     StageDeserialization,
			Message:                   err.Error(),
			ServiceResponseStatusCode: 0,
		}
	}

	return o, nil
}

func cleanRawResponse(s string) string {
	s = strings.TrimSpace(s)
	s = strings.TrimSuffix(s, ",")
	if s[:1] != "{" {
		s = "{" + s
	}
	if s[len(s)-1:len(s)] != "}" {
		s += "}"
	}
	return s
}

// REFACTOR: take to a dedicated helper function.
func (c *clientOpenAI) requestHandler(ctx context.Context, payload openAIRequest) ([]byte, error) {
	var w bytes.Buffer
	err := json.NewEncoder(&w).Encode(payload)
	if err != nil {
		return nil, Error{
			Service:                   ServiceOpenAI,
			Stage:                     StageSerialization,
			Message:                   err.Error(),
			ServiceResponseStatusCode: 0,
		}
	}

	req, _ := http.NewRequestWithContext(ctx, http.MethodPost, c.baseURL+"completions", &w)

	c.setHeader(req)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, Error{
			Service:                   ServiceOpenAI,
			Stage:                     StageRequest,
			Message:                   err.Error(),
			ServiceResponseStatusCode: 0,
		}
	}

	if resp.StatusCode > 209 {
		o := Error{
			Service:                   ServiceOpenAI,
			Stage:                     StageResponse,
			Message:                   "error status code: " + strconv.Itoa(resp.StatusCode),
			ServiceResponseStatusCode: resp.StatusCode,
		}

		var e openAIErrorResponse
		if err := json.NewDecoder(resp.Body).Decode(&e); err == nil {
			if v := e.Error; v != nil {
				o.Message = v.Message
			}
		}

		return nil, o
	}

	buf, err := io.ReadAll(resp.Body)
	defer func() { _ = resp.Body.Close() }()
	if err != nil {
		return nil, Error{
			Service:                   ServiceOpenAI,
			Stage:                     StageDeserialization,
			Message:                   err.Error(),
			ServiceResponseStatusCode: 0,
		}
	}

	return buf, nil
}

type openAIErrorResponse struct {
	Error *struct {
		Code    *int    `json:"code,omitempty"`
		Message string  `json:"message"`
		Param   *string `json:"param,omitempty"`
		Type    string  `json:"type"`
	} `json:"error,omitempty"`
}
